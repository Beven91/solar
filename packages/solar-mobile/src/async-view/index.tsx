/**
 * 名称：动态载入组件
 * 日期：2018-01-19
 * 描述：用于支持异步加载组件 （主要用于代码拆分，异步模块组件)
 *
 * AsyncView:
 *
 * 用例：
 *      <AsyncView component={getComponent} visible={this.state.visible}  >
 *      ..children... //需要作为getComponent组件子组件的内容
 *      </AsyncView>
 */
import React from 'react';
import Preload from '../preload';

const isFunction = (handler: any) => typeof handler === 'function';

type AsyncLoadable = (() => Promise<any>);
type AsyncComponent = React.ClassType<any, any, any> | AsyncLoadable;

export interface AsyncViewProps {
  // 异步组件获取函数 需要返回一个promise
  component: AsyncComponent,
  // 异步组件是否可见 当visible为false是，不会执行异步记载，当为true时进行加载
  visible?: boolean,
}

export interface AsyncViewState {
  // 组件是否在加载后立即渲染
  visible: boolean,
  // 需要渲染的组件
  requiredComponent: React.ReactElement<any>
}

export default class AsyncView extends React.Component<AsyncViewProps, AsyncViewState> {
  /**
   * 默认属性
   */
  static defaultProps = {
    visible: false,
  }

  private asyncComponent: AsyncComponent

  /**
   * 组件构造函数
   * @param props {Object} 组件props
   */
  constructor(props: AsyncViewProps) {
    super(props);
    this.asyncComponent = props.component;
    // 设置默认state
    this.state = {
      // 组件是否在加载后立即渲染
      visible: props.visible != false,
      // 需要渲染的组件
      requiredComponent: this.isAsyncComponent ? null : props.component as React.ReactElement<any>,
    };
  }

  /**
   * 判断当前props.component 是否为一个异步加载组件
   * 而不是纯粹的React组件
   */
  get isAsyncComponent() {
    const asyncComponent = this.asyncComponent as any;
    if (asyncComponent.prototype instanceof React.Component) {
      return false;
    } else if (isFunction(asyncComponent) && asyncComponent.length > 0) {
      // 如果有参数，则认为时一个函数组件
      return false;
    }
    return isFunction(asyncComponent);
  }

  /**
   * 组件第一次渲染完成时，执行一次异步获取
   */
  componentDidMount() {
    if (this.state.visible) {
      this.asyncRequireComponent();
    }
  }

  /**
   * 当路由切换时
   * @param nextProps {Object} 新的props
   */
  componentWillReceiveProps(nextProps: AsyncViewProps) {
    this.asyncComponent = nextProps.component;
    this.setState({ requiredComponent: null, visible: nextProps.visible });
    if (nextProps.visible) {
      this.asyncRequireComponent();
    }
  }

  /**
   * 仅在requiredComponent变更时，才刷新组件内容
   * @param nextProps {Object} 新的props
   * @param nextState {Object} 新的state
   * @returns {Boolean}
   */
  shouldComponentUpdate(nextProps: AsyncViewProps, nextState: AsyncViewState) {
    const requiredComponent = nextState.requiredComponent as any;
    return requiredComponent && requiredComponent.prototype instanceof React.Component;
  }

  /**
   * 尝试异步方式获取要显示的组件
   */
  asyncRequireComponent() {
    if (this.isAsyncComponent) {
      this.onProcessing();
      const asyncComponent = this.asyncComponent as AsyncLoadable;
      // 异步加载组件，然后进行渲染
      asyncComponent().then((m: any) => {
        const requiredComponent = (m.default || m);
        this.setState({ requiredComponent });
        this.onProcessed();
      });
    } else if (this.asyncComponent !== this.state.requiredComponent) {
      // 如果传入的是同步组件
      const requiredComponent = this.asyncComponent as React.ReactElement<any>;
      this.setState({ requiredComponent });
    }
  }

  /**
   * 异步组件请求中，可以在次函数中添加loaing效果
   */
  onProcessing() {
    // Preload.showLoading();
  }

  /**
   * 异步组件请求完毕
   */
  onProcessed() {
    Preload.closeLoading();
  }

  /**
   * 组件渲染
   */
  render() {
    const { state } = this;
    const Component = state.requiredComponent as any;
    if (Component && state.visible) {
      return (
        <div className="async-view">
          <Component {...this.props} />
        </div>
      );
    }
    return '';
  }
}
