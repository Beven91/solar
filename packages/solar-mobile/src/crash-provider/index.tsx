/**
 * @module CrashProvider
 * @description 渲染异常容器，当子元素渲染异常，则会显示一个异常页面
 */
import React, { ReactElement } from 'react';
import Crash from '../crash';

export interface CrashError {
  error: string
}

export interface CrashProviderProps {
  // 返回异常显示内容
  view?: (error: CrashError) => ReactElement
  // 当前error是否可见
  visible?: boolean
  // 强制指定异常信息
  error?: string
  // 自定义Error组件属性
  errorProps?: object
}

interface CrashProviderState {
  error: string
}

export default class CrashProvider extends React.Component<CrashProviderProps, CrashProviderState> {
  // 默认属性
  static defaultProps = {
    view: () => <Crash />,
    visible: false,
  }

  /**
   *当子元素出现异常
   * @param {Error} e 错误对象
   */
  static getDerivedStateFromError() {
    return {
      error: '渲染异常',
    };
  }

  state = {
    error: '',
  }

  componentDidCatch() {
    return true;
  }

  // 渲染一个异常状态视图
  renderError() {
    const { error } = this.state;
    const { view } = this.props;
    return view({ error: (error || this.props.error) as string });
  }

  // 渲染组件
  render() {
    if (this.state.error || this.props.visible) {
      return this.renderError();
    }
    return React.Children.only(this.props.children);
  }
}
