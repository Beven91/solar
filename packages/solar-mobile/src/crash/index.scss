.crash-component {
  height: 100%;
  background-color: #f2f2f2;
  overflow: hidden;
  text-align: center;
  color:#616161;

  .logo {
    width: 100px;
    margin: 180px auto 26px;
    display: block;
  }

  .crash-title {
    font-size: 16px;
    font-weight: 700;
  }

  .crash-description{
    font-size:13px;
    color:#9e9e9e;
    margin-top:18px;
  }

  .crash-button{
    margin-top:20px;
    font-size:14px;
    display: inline-block;
    min-width: 70px;
    padding:10px 15px;
    border:1px solid #bdbdbd;
  }
}