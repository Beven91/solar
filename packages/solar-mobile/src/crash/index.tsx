/**
 * @name 异常处理组件
 * @description 用于显示一个异常视图
 */
import './index.scss';
import React from 'react';
import defaultIcon from './images/error.svg';

export interface CrashProps {
  /**
   * 要显示的消息
   */
  message?: string
  // 标题
  title?: string
  // 按钮标题
  btnText?: string
  // 自定义图片
  icon?: string
  // 自定义样式名
  className?: string
  // 自定义样式
  style?: object;
  // 点击按钮事件
  onClick?: () => void
}

export default class Crash extends React.Component<CrashProps, {}> {
  // 默认属性
  static defaultProps = {
    onClick: () => history.back(),
    className: '',
    btnText: '返回',
    title: '页面竟然崩溃了',
    message: '大侠别急，您可点击以返回试试~',
    icon: defaultIcon,
  }

  // 渲染组件
  render() {
    const { btnText, title, message, className, style, onClick, icon } = this.props;
    return (
      <div style={style} className={`crash-component  ${className}`}>
        <img className="logo" src={icon || ''} />
        < div className="crash-title" > {(title)}</div>
        {
          !message ? null : (
            <div className="crash-description" >
              {message}
            </div>
          )
        }
        { onClick ? <div className="crash-button" onClick={onClick} > {btnText} </div> : null}
        {this.props.children}
      </div>
    );
  }
}
