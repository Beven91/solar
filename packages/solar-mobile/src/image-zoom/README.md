# image-zoom

图片预览缩放组件（点击容器中的图片，可弹出浮层，可以对图片进行打开查看)

- [用例](/__tests__/image-zoom/index.js)

```js
import React from 'react';
import { ImageZoom } from 'ireeder-ui-mobile';

export default class App extends React.Component {
  render() {
    return (
      <div style={styles.container}>
        <ImageZoom>
          <div>一下图片点击可放大哦</div>
          <img style={styles.img} 
              src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100" 
              data-zoom="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg" 
          />
          <img style={styles.img} src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100" />
           <img style={styles.img} src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100" />
            <img style={styles.img} src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100" />
        </ImageZoom>
      </div>
    )
  }
}
```

### 参数

<Highlight lang="language-javascript" :code="$page.componentProps" />
