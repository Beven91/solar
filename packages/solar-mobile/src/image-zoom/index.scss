.image-zoom-container{
  height:100%;
  width:100%;
  position: fixed;
  z-index:1000;
  display:flex;
  top:0px;
  bottom:0px;
  left:0px;
  right:0px;
  justify-content:center;
  align-items: center;
  background-color:#000;
  display:none;
  .image{
    width:100%;
    user-select: none;
  }
}
.image-zoom-container.show{
  display:flex;
}

.pinch-zoom-container{
  width: 100%;
}