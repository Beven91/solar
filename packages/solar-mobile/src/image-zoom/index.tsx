/**
 * @name ImageZoom 图片放大容器组件
 * @date 2018-05-02
 * @description
 *       支持容器下，指定类型图片标签在点击
 */
import './index.scss';
import '../polyfill/requestAnimation';
import PinchZoom from 'pinch-zoom-js';
import React from 'react';

export default class ImageZoom extends React.Component {
  // 点击事件，在点击图片时,进行放大
  onPinchZoom = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const target = e.target as HTMLElement;
    const src = target.getAttribute('src');
    const bigSrc = target.getAttribute('data-zoom');
    const url = bigSrc || src;
    if (url) {
      const html = `
        <div class="image-zoom-container show">
          <div>
            <img src="${url}" class="image"/>
          </div>
        </div>
      `;
      const div = document.createElement('div');
      div.innerHTML = html;
      div.addEventListener('click', function() {
        document.body.removeChild(div);
      });
      document.body.appendChild(div);
      new PinchZoom(div.children[0].children[0] as HTMLElement);
    }
  }

  // 渲染组件
  render() {
    return (
      <div onClick={this.onPinchZoom} {...this.props}>
        {this.props.children}
      </div>
    );
  }
}
