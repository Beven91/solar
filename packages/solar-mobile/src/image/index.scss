.image-dynamic {
  display: inline-block;
}

.image-dynamic.dynamic-hidden {
  visibility: hidden;
}