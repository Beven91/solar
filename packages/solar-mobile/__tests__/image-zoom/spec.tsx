import React from 'react';
import PinchZoom from 'pinch-zoom-js';
import ImageZoom from '../../src/image-zoom';
import AnimatePolyfill from '../../src/polyfill/requestAnimation';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('pinch-zoom-js');

describe('ImageZoom', () => {
  it('zoom full parameter', () => {
    const tree = mount(
      <ImageZoom>
        <img
          className="with-zoom"
          src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100"
          data-zoom="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg"
        />
      </ImageZoom>
    );

    // 断言：对比默认的渲染是否一致
    expect(toJson(tree)).toMatchSnapshot();

    let zoomContainer = null as HTMLDivElement;

    const mock = PinchZoom as any as jest.MockInstance<any, any>;

    mock.mockImplementation((div) => zoomContainer = div);

    // 模拟点击放大
    tree.find('.with-zoom').simulate('click');
    // 断言：是否调用了PinchZoom插件
    expect(PinchZoom).toHaveBeenCalledTimes(1);
    expect(PinchZoom).toBeCalledWith(zoomContainer);
    // 断言：通过PinchZoom渲染后的body的渲染是否一致
    expect(document.body).toMatchSnapshot();

    // 测试点击浮层 关闭放大弹窗
    zoomContainer.click();
    // 断言：渲染
    expect(document.body).toMatchSnapshot();
  });


  it('zoom img without data-zoom', () => {
    document.body.innerHTML = '';
    const tree = mount(
      <ImageZoom>
        <img className="default" src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100" />
      </ImageZoom>
    );

    // 断言：对比默认的渲染是否一致
    expect(toJson(tree)).toMatchSnapshot();

    // 模拟点击放大
    tree.find('.default').simulate('click');
    // 断言：通过PinchZoom渲染后的body的渲染是否一致
    expect(document.body).toMatchSnapshot();
  });


  it('zoom img no src', () => {
    document.body.innerHTML = '';
    const tree = mount(
      <ImageZoom>
        <img className="default" />
      </ImageZoom>
    );

    // 断言：对比默认的渲染是否一致
    expect(toJson(tree)).toMatchSnapshot();

    // 模拟点击放大
    tree.find('.default').simulate('click');
    // 断言：通过PinchZoom渲染后的body的渲染是否一致
    expect(document.body).toMatchSnapshot();
  });


  it('polyfill', () => {
    const fn = jest.fn();
    const fn2 = jest.fn();
    jest.useFakeTimers();

    AnimatePolyfill.requestAnimationFrame(fn);
    // 时间推进1秒
    jest.advanceTimersByTime(1000);
    // 断言：requestAnimationFrame 的fn被执行一次
    expect(fn).toHaveBeenCalledTimes(1);

    const id = AnimatePolyfill.requestAnimationFrame(fn2);
    // 取消 requestAnimationFrame
    AnimatePolyfill.cancelAnimationFrame(id);
    // 时间推进1秒
    jest.advanceTimersByTime(1000);
    // 断言：requestAnimationFrame 的fn不会被执行，因为：被取消
    expect(fn2).toHaveBeenCalledTimes(0);

    // 清除require模块缓存
    const fn3 = jest.fn();
    jest.resetModules();
    window.requestAnimationFrame = null;
    window.cancelAnimationFrame = null;
    require('../../src/image-zoom').default;
    window.requestAnimationFrame(fn3);
    // 时间推进1秒
    jest.advanceTimersByTime(1000);
    // 断言：requestAnimationFrame 的fn被执行一次
    expect(fn3).toHaveBeenCalledTimes(1);
  });
});

