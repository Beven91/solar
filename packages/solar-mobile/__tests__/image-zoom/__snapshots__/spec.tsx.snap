// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageZoom zoom full parameter 1`] = `
<ImageZoom>
  <div
    onClick={[Function]}
  >
    <img
      className="with-zoom"
      data-zoom="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg"
      src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100"
    />
  </div>
</ImageZoom>
`;

exports[`ImageZoom zoom full parameter 2`] = `
<body>
  <div>
    
        
    <div
      class="image-zoom-container show"
    >
      
          
      <div>
        
            
        <img
          class="image"
          src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg"
        />
        
          
      </div>
      
        
    </div>
    
      
  </div>
</body>
`;

exports[`ImageZoom zoom full parameter 3`] = `<body />`;

exports[`ImageZoom zoom img no src 1`] = `
<ImageZoom>
  <div
    onClick={[Function]}
  >
    <img
      className="default"
    />
  </div>
</ImageZoom>
`;

exports[`ImageZoom zoom img no src 2`] = `<body />`;

exports[`ImageZoom zoom img without data-zoom 1`] = `
<ImageZoom>
  <div
    onClick={[Function]}
  >
    <img
      className="default"
      src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100"
    />
  </div>
</ImageZoom>
`;

exports[`ImageZoom zoom img without data-zoom 2`] = `
<body>
  <div>
    
        
    <div
      class="image-zoom-container show"
    >
      
          
      <div>
        
            
        <img
          class="image"
          src="https://static.solar.cn/T16lVbB4J41RCvBVdK.jpg?img=/rs,w_100"
        />
        
          
      </div>
      
        
    </div>
    
      
  </div>
</body>
`;
